File: v8-core/src/system_proving/receipt.rs
//! System Proving Ground Receipts (D-147, D-149, M3).
//!
//! Sealed receipts documenting end-to-end stress proving ground executions.

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use crate::system_proving::attribution::FailureAttributionBreakdown;
use crate::system_proving::metrics::SystemRobustnessVector;

/// Final receipt emitted by the System Proving Ground.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SystemProvingGroundReceipt {
    pub receipt_id: String,
    pub world_id: String,
    pub policy_id: String,
    pub total_trades: usize,
    pub total_campaigns: usize,
    pub metrics: SystemRobustnessVector,
    pub attribution: FailureAttributionBreakdown,
    pub exercises_full_pipeline: bool,
    pub evaluated_at_timestamp_ns: u64,
    pub receipt_digest: String,
}

impl SystemProvingGroundReceipt {
    pub fn new(
        world_id: String,
        policy_id: String,
        total_trades: usize,
        total_campaigns: usize,
        metrics: SystemRobustnessVector,
        attribution: FailureAttributionBreakdown,
        exercises_full_pipeline: bool,
        timestamp_ns: u64,
    ) -> Self {
        let mut hasher = Sha256::new();
        hasher.update(world_id.as_bytes());
        hasher.update(policy_id.as_bytes());
        hasher.update(&(total_trades as u64).to_le_bytes());
        hasher.update(&(total_campaigns as u64).to_le_bytes());
        hasher.update(&timestamp_ns.to_le_bytes());
        let digest = format!("{:x}", hasher.finalize());
        let receipt_id = format!("spg-receipt-{}", &digest[..16]);

        Self {
            receipt_id,
            world_id,
            policy_id,
            total_trades,
            total_campaigns,
            metrics,
            attribution,
            exercises_full_pipeline,
            evaluated_at_timestamp_ns: timestamp_ns,
            receipt_digest: digest,
        }
    }
}

File: v8-next/src/v8_next/system_proving/receipt.py
"""System proving ground receipt — port of v8-core/src/system_proving/receipt.rs.

The digest covers the world/policy identity, the trade and campaign counts and the timestamp,
exactly as the Rust receipt does; the receipt id is ``spg-receipt-<first 16 hex>``.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass

from v8_next.system_proving.attribution import FailureAttributionBreakdown
from v8_next.system_proving.metrics import SystemRobustnessVector


@dataclass(frozen=True)
class SystemProvingGroundReceipt:
    receipt_id: str
    world_id: str
    policy_id: str
    total_trades: int
    total_campaigns: int
    metrics: SystemRobustnessVector
    attribution: FailureAttributionBreakdown
    exercises_full_pipeline: bool
    evaluated_at_timestamp_ns: int
    receipt_digest: str

    @classmethod
    def new(
        cls,
        *,
        world_id: str,
        policy_id: str,
        total_trades: int,
        total_campaigns: int,
        metrics: SystemRobustnessVector,
        attribution: FailureAttributionBreakdown,
        exercises_full_pipeline: bool,
        timestamp_ns: int,
    ) -> SystemProvingGroundReceipt:
        digest = hashlib.sha256(
            (
                f"{world_id}|{policy_id}|{total_trades}|{total_campaigns}|{timestamp_ns}"
            ).encode()
        ).hexdigest()
        return cls(
            receipt_id=f"spg-receipt-{digest[:16]}",
            world_id=world_id,
            policy_id=policy_id,
            total_trades=total_trades,
            total_campaigns=total_campaigns,
            metrics=metrics,
            attribution=attribution,
            exercises_full_pipeline=exercises_full_pipeline,
            evaluated_at_timestamp_ns=timestamp_ns,
            receipt_digest=digest,
        )

    def as_dict(self) -> dict[str, object]:
        return {
            "receipt_id": self.receipt_id,
            "world_id": self.world_id,
            "policy_id": self.policy_id,
            "total_trades": self.total_trades,
            "total_campaigns": self.total_campaigns,
            "metrics": self.metrics.as_dict(),
            "attribution": self.attribution.as_dict(),
            "exercises_full_pipeline": self.exercises_full_pipeline,
            "evaluated_at_timestamp_ns": self.evaluated_at_timestamp_ns,
            "receipt_digest": self.receipt_digest,
        }
